File: ts/2015/day1/day1.ts
export const getFloor = (instruction: string, currentFloor = 0): number => {
  const parens = instruction.split('');

  let floor = currentFloor;

  parens.forEach(paren => {
    if (paren === '(') floor++;
    if (paren === ')') floor--;
  })

  return floor;
}

export const getPositionToFirstEnterBasement = (instruction: string): number => {
  let position = 0;

  for (let i = 1; i <= instruction.length; i++) {
    const floor = getFloor(instruction.substring(0, i));

    if (floor === -1) {
      position = i;
      break;
    }
  }

  return position;
}

export const getPositionToFirstEnterBasementV2 = (instruction: string): number => {
  let position = 0;
  let currentFloor = 0;

  for (let i = 0; i <= instruction.length; i++) {
    const floor = getFloor(instruction[i], currentFloor)
    
    if (floor === -1) {
      position = i + 1;
      break;
    }

    currentFloor = floor
  }
  
  return position;
}

export const getPositionToFirstEnterBasementV3 = (instruction: string): number => {

  const parens = instruction.split('')

  let currentFloor = 0;
  let position = 0;
  
  parens.forEach((paren, index) => {
    const floor = getFloor(paren, currentFloor)
    
    if (currentFloor !== -1) {
      currentFloor = floor
      position = index + 1
    }
  })

  return position;
}

File: go/2015/day1/day1.go
package day1

import "strings"

func GetFloor(instruction string, currentFloor int) int {
	parens := strings.Split(instruction, "")

	floor := currentFloor

	for _, paren := range parens {
		if paren == "(" {
			floor++
		}

		if paren == ")" {
			floor--
		}
	}

	return floor
}

// Index started with 0 and only up to len - 1, since we are accessing each character of the instruction
func GetPositionToFirstEnterBasement(instruction string) int {
	position := 0
	currentFloor := 0

	for i := 0; i < len(instruction); i++ {
		paren := string(instruction[i])
		floor := GetFloor(paren, currentFloor)

		if floor == -1 {
			position = i + 1
			break
		}

		currentFloor = floor
	}

	return position
}

// Index started with 1 up to len, since we are using it as exclusive upper bound
func GetPositionToFirstEnterBasementV2(instruction string) int {
	position := 0

	for i := 1; i <= len(instruction); i++ {
		parens := string(instruction[:i])

		floor := GetFloor(parens, 0)

		if floor == -1 {
			position = i
			break
		}
	}

	return position
}
